File: go-leetcode/backtracking/17.电话号码的字母组合.go
package leetcode

func letterCombinations(digits string) []string {
	if len(digits) == 0 {
		return []string{}
	}

	// 数字到字母的映射
	phoneMap := map[byte]string{
		'2': "abc",
		'3': "def",
		'4': "ghi",
		'5': "jkl",
		'6': "mno",
		'7': "pqrs",
		'8': "tuv",
		'9': "wxyz",
	}

	res := []string{}
	var backtrack func(index int, path string)

	backtrack = func(index int, path string) {
		if len(path) == len(digits) {
			res = append(res, path)
			return
		}
		letters := phoneMap[digits[index]]
		for i := 0; i < len(letters); i++ {
			backtrack(index+1, path+string(letters[i]))
		}
	}

	backtrack(0, "")
	return res
}

File: python-leetcode/backtracking/17.电话号码的字母组合.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        phone_map: dict[str, str] = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        res: List[str] = []

        def backtrack(index: int, path: str):
            if len(path) == len(digits):
                res.append(path)
                return

            letters = phone_map[digits[index]]
            for char in letters:
                backtrack(index + 1, path + char)

        backtrack(0, "")
        return res
